WalletWasabi/WabiSabi/Client/ArenaClient.cs

```diff
@@ -0,0 +1,99 @@
+using WalletWasabi.Crypto.ZeroKnowledge;
+using WalletWasabi.Helpers;
+using WalletWasabi.WabiSabi.Backend.PostRequests;
+using WalletWasabi.WabiSabi.Crypto;
+using WalletWasabi.WabiSabi.Models;
+
+namespace WalletWasabi.WabiSabi.Client
+{
+	public class ArenaClient
+	{
+		public ArenaClient(WabiSabiClient amountCredentialClient, WabiSabiClient weightCredentialClient, IRequestHandler requestHandler)
+		{
+			AmountCredentialClient = amountCredentialClient;
+			WeightCredentialClient = weightCredentialClient;
+			RequestHandler = requestHandler;
+		}
+
+		public WabiSabiClient AmountCredentialClient { get; }
+		public WabiSabiClient WeightCredentialClient { get; }
+		public IRequestHandler RequestHandler { get; }
+
+		public ValueTask<Guid> RegisterInputAsync(Money amount, OutPoint outPoint, Key key, Guid roundId, uint256 roundHash) =>
+			RegisterInputAsync(
+				new[] { amount },
+				new[] { outPoint },
+				new[] { key },
+				roundId,
+				roundHash);
+
+		public async ValueTask<Guid> RegisterInputAsync(
+			IEnumerable<Money> amounts,
+			IEnumerable<OutPoint> outPoints,
+			IEnumerable<Key> keys,
+			Guid roundId,
+			uint256 roundHash)
+		{
+			static byte[] GenerateOwnershipProof(Key key, uint256 roundHash) => OwnershipProof.GenerateCoinJoinInputProof(
+				key,
+				new CoinJoinInputCommitmentData("CoinJoinCoordinatorIdentifier", roundHash)).ToBytes();
+
+			var registrableInputs = outPoints
+				.Zip(keys, (outPoint, key) => (outPoint, key))
+				.Select(x => new InputRoundSignaturePair(x.outPoint, GenerateOwnershipProof(x.key, roundHash)));
+
+			var (zeroAmountCredentialRequest, zeroAmountCredentialResponseValidation) = AmountCredentialClient.CreateRequestForZeroAmount();
+			var (zeroWeightCredentialRequest, zeroWeightCredentialResponseValidation) = WeightCredentialClient.CreateRequestForZeroAmount();
+
+			var inputRegistrationResponse = await RequestHandler.RegisterInputAsync(
+				new InputsRegistrationRequest(
+					roundId,
+					registrableInputs,
+					zeroAmountCredentialRequest,
+					zeroWeightCredentialRequest)).ConfigureAwait(false);
+
+			AmountCredentialClient.HandleResponse(inputRegistrationResponse.AmountCredentials, zeroAmountCredentialResponseValidation);
+			WeightCredentialClient.HandleResponse(inputRegistrationResponse.WeightCredentials, zeroWeightCredentialResponseValidation);
+
+			return inputRegistrationResponse.AliceId;
+		}
+
+		public async Task ConfirmConnectionAsync(Guid roundId, Guid aliceId, IEnumerable<Credential> credentialsToPresent, IEnumerable<Money> newAmount)
+		{
+			var (realAmountCredentialRequest, realAmountCredentialResponseValidation) = AmountCredentialClient.CreateRequest(
+				newAmount.Select(x => x.Satoshi),
+				credentialsToPresent);
+
+			var inputsCount = 1; // ?
+			var (realWeightCredentialRequest, realWeightCredentialResponseValidation) = WeightCredentialClient.CreateRequest(
+				new[] { 1_000L - (inputsCount) * 4 * Constants.P2wpkhInputVirtualSize },
```

Comment:
can be refactored into `inputRegistrationWeight` parameter or something like that, connection confirmation should not care anymore how many prevouts are used or what type they are

Reply:
Done.